### PR TITLE
feat: add on-demand full-state checkpointing for OSFT training resumption

### PR DESCRIPTION
**Pull request by RobotSail**
## Summary

Implements signal-driven full-state checkpointing that enables exact training resumption for OSFT models in Kubernetes/OpenShift/SLURM environments.

- **Signal handling**: Catches 7 termination signals (SIGTERM, SIGINT, SIGUSR1/2, SIGXCPU, SIGHUP, SIGQUIT) covering Kubernetes, SLURM, PBS, and LSF schedulers
- **Trigger mechanism**: Signal handler writes to `/dev/shm` trigger file; workers poll at batch boundaries and coordinate via `all_reduce(MAX)` for distributed consensus
- **DCP sharded save**: Each rank saves its own shard via `torch.distributed.checkpoint` — no gathering to rank 0, minimal communication overhead
- **Exact resumption**: Saves complete training state (model OSFT factors, optimizer, LR scheduler, per-rank RNG states, training counters, checkpointer state) enabling bit-identical optimization trajectories after resume
- **OSFT-aware resume**: On restart, the model structure is initialized normally (with SVD computation for meta tensor materialization), then all parameters are overwritten with checkpoint values via DCP in-place load

### New files
- `src/mini_trainer/full_state_checkpoint.py` — `FullStateCheckpointer` class
- `tests/test_full_state_checkpoint.py` — 17 unit tests
- `regression_tests/test_full_state_checkpoint_fidelity.py` — Trajectory fidelity test
- `regression_tests/run_multi_node_fidelity_test.sh` — Multi-node simulation test (2 nodes x 4 GPUs)
- `regression_tests/test_full_state_checkpoint_signal.sh` — Signal-triggered end-to-end test

### Modified files
- `src/mini_trainer/training_types.py` — Added `on_demand_checkpointing` and `resume_from_full_state_checkpoint` fields
- `src/mini_trainer/setup_model_for_training.py` — Resume path in `finalize_model_initialization()` that loads from DCP instead of computing SVD
- `src/mini_trainer/train.py` — Training loop integration with save trigger, resume state restoration, and step-based batch skipping

## Test plan

- [x] 17 unit tests for signal handling, trigger detection, metadata/RNG roundtrip, load/save mechanics
- [x] 276 existing unit tests still pass (0 regressions)
- [x] Optimization trajectory fidelity test: 20-step baseline vs checkpoint-at-10/resume proves **bit-identical** loss, grad_norm, lr, and parameter hashes across all steps
- [x] Multi-node fidelity test: Same fidelity verification across 2 simulated nodes (8 GPUs total)
- [x] Signal-triggered end-to-end test: Training with `--on-demand-checkpointing` in infinite mode, trigger file written, all ranks checkpoint and exit cleanly, resume continues from correct step
- [x] Ruff lint and format checks pass

## Summary by CodeRabbit

* **New Features**
  * On‑demand, signal‑triggered full‑state checkpointing for distributed training.
  * Resume from full‑state checkpoints restoring model shards, optimizer, RNG, scheduler and training progress; new CLI flags to enable checkpointing and resume.

* **Documentation**
  * Added a comprehensive design spec for on‑demand full‑state checkpointing.

* **Tests**
  * Added unit tests and multiple regression harnesses (multi‑node, signal‑trigger, per‑model matrix) validating fidelity and resume workflows.

### CODE REVIEW

File: src/mini_trainer/setup_model_for_training.py [hunk 1]
```diff
@@ -571,15 +574,62 @@ def finalize_model_initialization(model: torch.nn.Module, context: ModelInitiali
...
+                        mod._buffers[buf_name] = new_dt
+                        materialized += 1
+            log_rank_0(f"   ✓ Materialized {materialized} meta DTensors")
+
+            # Step 2: Load checkpoint values via DCP in-place
+            log_rank_0("🔄 [OSFT] Loading checkpoint state via DCP")
+            checkpoint_dir = Path(context.resume_from_checkpoint)
+            dcp_dir = str(checkpoint_dir / "distributed")
+            dcp.load({"model": model.state_dict()}, checkpoint_id=dcp_dir)
+            log_rank_0("   ✓ Model state loaded from checkpoint")
```
[the remark this answers is not in the document]
Comment (RobotSail) [reply src/mini_trainer/setup_model_for_training.py]: We intentionally use `model.state_dict()` here instead of the canonical `get_model_state_dict`/`set_model_state_dict` pattern. During development we found that `set_model_state_dict` corrupts FSDP2's internal sharding metadata when called after the model has already been materialized via our DTensor replacement approach, causing NCCL hangs on the first forward pass. The `model.state_dict()` call returns DTensor references that DCP can load into in-place without disturbing the sharding state. This was validated across 5 model architectures (Llama, Qwen2, Granite 8B, Phi4, Granite-MoE) with both liger on and off.

**Comment by RobotSail** [comment]
@claude-code do you know why these tests are failing? Can you please fix them?

**Review by Maxusmusti — APPROVED** [review]
LGTM
